pkg/validator/validator.go

```diff
@@ -15,12 +15,141 @@
+	IssueIds []string `json:"issueIds"`
+	Jqls     []string `json:"jqls"`
+}
+
+func NewValidator(url string) (*Validator, error) {
+	return &Validator{
+		baseURL:    url,
+		httpClient: &http.Client{Timeout: 10 * time.Second},
+	}, nil
+}
+
+func (v *Validator) MatchIssue(ctx context.Context, issueKey, jql, jiraAccount, apiToken string) (*MatchResult, error) {
+	issue, err := v.jiraIssue(ctx, issueKey, jiraAccount, apiToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get jira issue %s: %w", issueKey, err)
+	}
+
+	result, err := v.matchJQL(ctx, issue, jql, jiraAccount, apiToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate jira issue %s: %w", issueKey, err)
+	}
+	return result, nil
+}
+
+func (v *Validator) jiraIssue(ctx context.Context, issueIDOrKey, jiraAccount, apiToken string) (*JiraIssue, error) {
+	// Construct get issue api.
+	// https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issues/#api-rest-api-3-issue-issueidorkey-get
+	u, err := url.Parse(fmt.Sprintf("%s/issue/%s?fields=key,id", v.baseURL, issueIDOrKey))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse getIssue url: %w", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct request: %w", err)
+	}
+
+	header := map[string]string{"Accept": "application/json"}
+
+	var jiraIssue JiraIssue
+	if err := v.makeRequest(req, jiraAccount, apiToken, header, &jiraIssue); err != nil {
+		return nil, err
+	}
+
+	return &jiraIssue, nil
+}
+
+func (v *Validator) matchJQL(ctx context.Context, issue *JiraIssue, jql, jiraAccount, apiToken string) (*MatchResult, error) {
+	// Construct match api.
+	// https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-search/#api-rest-api-3-jql-match-post
+
+	u, err := url.Parse(fmt.Sprintf("%s/jql/match", v.baseURL))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse getIssue url: %w", err)
+	}
+
+	// Create the request body.
+	data := matchData{
+		IssueIds: []string{issue.ID},
+		Jqls:     []string{jql},
+	}
+	body, err := json.Marshal(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct request body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), bytes.NewBuffer(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct request: %w", err)
+	}
+
+	header := map[string]string{"Accept": "application/json", "Content-Type": "application/json"}
+
+	var result MatchResult
+	if err := v.makeRequest(req, jiraAccount, apiToken, header, &result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// Make request and store the data in the value pointed by respVal.
+func (v *Validator) makeRequest(req *http.Request, jiraAccount, apiToken string, header map[string]string, respVal any) error {
+	req.SetBasicAuth(jiraAccount, apiToken)
+	for k, val := range header {
+		req.Header.Set(k, val)
```

Comment:
This is a little weird, especially since it's not documented. This is modifying the header state of the provided request. Meaning the other methods now have added headers. It's not a huge deal, but it's somewhat unexpected.

Reply:
Can you just set the header in the request (before calling) instead of having the header as an additional parameter?

Reply:
I think I misunderstood suggestions from https://github.com/abcxyz/jvs-plugin-jira/pull/9#discussion_r1225866481. Moved the SetHeader and keep the SetBasicAuth in makeRequest now. Let me know what you think